platform/darwin/src/MGLMapSnapshotter.h

```diff
@@ -284,6 +288,78 @@ MGL_EXPORT
  */
 @property (nonatomic, readonly, getter=isLoading) BOOL loading;
 
+/**
+ The snapshotter’s delegate.
+ 
+ The delegate is responsible for responding to significant changes during the
+ snapshotting process, such as the style loading. Implement a delegate to
+ customize the style that is depicted by the snapshot.
+ 
+ You set the delegate after initializing the snapshotter but before receiving
+ the snapshot, typically before starting the snapshot. The snapshotter keeps a
+ weak reference to its delegate, so you must keep a strong reference to it to
+ ensure that your style customizations apply.
+ */
+@property (nonatomic, weak) id <MGLMapSnapshotterDelegate> delegate;
+
+/**
+ The style displayed in the resulting snapshot.
+ 
+ Unlike the `MGLMapSnapshotOptions.styleURL` property, this property is set to
+ an object that allows you to manipulate every aspect of the style locally.
+ 
+ This property is set to `nil` until the style finishes loading. If the style
+ has failed to load, this property is set to `nil`. Because the style loads
+ asynchronously, you should manipulate it in the
+ `-[MGLMapSnapshotterDelegate mapSnapshotter:didFinishLoadingStyle:]` method. It
+ is not possible to manipulate the style before it has finished loading.
+
+ @note The default styles provided by Mapbox contain sources and layers with
+    identifiers that will change over time. Applications that use APIs that
+    manipulate a style’s sources and layers must first set the style URL to an
+    explicitly versioned style using a convenience method like
+    `+[MGLStyle outdoorsStyleURLWithVersion:]` or a manually constructed
+    `NSURL`.
+ */
+@property (nonatomic, readonly, nullable) MGLStyle *style;
+
+@end
+
+/**
+ Optional methods about significant events when creating a snapshot using an
+ `MGLMapSnapshotter` object.
+ */
+@protocol MGLMapSnapshotterDelegate <NSObject>
+@optional
+
+/**
+ Tells the delegate that the snapshotter was unable to load data needed for
+ snapshotting the map.
+ 
+ This method may be called for a variety of reasons, including a network
+ connection failure or a failure to fetch the style from the server. You can use
+ the given error message to notify the user that map data is unavailable.
+ 
+ @param snapshotter The snapshotter that is unable to load the data.
+ @param error The reason the data could not be loaded.
+*/
+- (void)mapSnapshotterDidFail:(MGLMapSnapshotter *)snapshotter withError:(NSError *)error;
```

Comment:
Does this overlap with the error code passed to the `MGLMapSnapshotCompletionHandler`? I think there might be some scope for confusion here.

Reply:
So far this delegate method only gets called if the style fails to load. But I intentionally worded it more generically in case we need to support other failures in the future this way, particularly recoverable errors. A failure to load the style also causes the snapshot to fail, so an error still gets passed into the completion handler. This delegate method just gives additional information about the cause.

Reply:
So to confirm - in the case of an error where the dev implements this method - they would see the error here, and then finally in the completion block? Would it be the same error or this one would be more specific?

Reply:
Yes, they’d see both errors. This error would be more specific, though I wouldn’t make that part of the API contract since mbgl’s API contract doesn’t say what errors show up when.